internal/pkg/service/buffer/sink/tablesink/storage/repository/file_repo.go

```diff
@@ -294,130 +330,154 @@ func (r *FileRepository) rotateAllIn(rb rollback.Builder, now time.Time, parentK
-	// Similarly, slices in the SliceWriting state are switched to the SliceClosing state.
+	// Read opened files in the storage.FileWriting state.
+	// There can be a maximum of one old file in the storage.FileWriting state per each table sink.
+	// On rotation, opened files are switched to the storage.FileClosing state.
 	var openedFiles []storage.File
-	var openedSlices []storage.Slice
-	atomicOp.
-		ReadOp(r.ListInState(parentKey, storage.FileWriting).WithAllTo(&openedFiles)).
-		ReadOp(r.all.slice.ListInState(parentKey, storage.SliceWriting).WithAllTo(&openedSlices))
+	atomicOp.ReadOp(r.ListInState(parentKey, storage.FileWriting).WithAllTo(&openedFiles))
 
-	// Close old files, open new files
-	atomicOp.WriteOrErr(func(ctx context.Context) (op.Op, error) {
-		txn := op.Txn(r.client)
+	// Read opened slices in the storage.SliceWriting state.
+	// On rotation, opened slices are switched to the storage.SliceClosing state.
+	var openedSlices []storage.Slice
+	atomicOp.ReadOp(r.all.slice.ListInState(parentKey, storage.SliceWriting).WithAllTo(&openedSlices))
 
-		// Group opened files by the sink
-		openedFilesPerSink := make(map[key.SinkKey][]storage.File)
+	// Group opened files by the sink
+	var openedFilesPerSink map[key.SinkKey][]storage.File
+	atomicOp.BeforeWrite(func(ctx context.Context) {
+		openedFilesPerSink = make(map[key.SinkKey][]storage.File)
 		for _, file := range openedFiles {
 			openedFilesPerSink[file.SinkKey] = append(openedFilesPerSink[file.SinkKey], file)
 		}
+	})
 
-		// Group opened slices by the file
-		openedSlicesPerSink := make(map[key.SinkKey][]storage.Slice)
+	// Group opened slices by the file
+	var openedSlicesPerSink map[key.SinkKey][]storage.Slice
+	atomicOp.BeforeWrite(func(ctx context.Context) {
+		openedSlicesPerSink = make(map[key.SinkKey][]storage.Slice)
 		for _, slice := range openedSlices {
 			openedSlicesPerSink[slice.SinkKey] = append(openedSlicesPerSink[slice.SinkKey], slice)
 		}
+	})
 
-		// Open a new file in each sink
+	// Close old files, open new files
+	atomicOp.WriteOrErr(func(ctx context.Context) (op.Op, error) {
+		txn := op.Txn(r.client)
 		errs := errors.NewMultiError()
+
+		// Open a new file in each sink
 		for _, sink := range sinks {
-			// File should be opened only for the table sinks
-			if sink.Type != definition.SinkTypeTable {
-				continue
+			sinkTxn, err := r.rotateSink(ctx, rotateSinkContext{
+				Now:             now,
+				OpenNew:         openNew,
+				Sink:            sink,
+				Volumes:         volumes,
+				OpenedFiles:     openedFilesPerSink[sink.SinkKey],
+				OpenedSlices:    openedSlicesPerSink[sink.SinkKey],
+				MaxUsedDiskSize: maxUsedDiskSpace[sink.SinkKey],
+				NewFileResource: fileResources[sink.SinkKey],
+			})
+
+			if err != nil {
+				errs.Append(err)
+			} else if sinkTxn != nil {
+				txn.Merge(sinkTxn.OnSucceeded(func(r *op.TxnResult[*storage.File]) {
+					if f := r.Result(); f != nil {
+						newFiles = append(newFiles, *f)
+					}
+				}))
 			}
```

Comment:
New implementation of the `for` loop.

Code from the `for` loop has been moved to the `rotateSink` method (see side-by-side diff).

I also thought about creating `rotateAllInContext`, but it didn't improve readability.